File: concrete-sample/src/main/java/com/jaynewstrom/concretesample/application/ConcreteSampleApplication.java
package com.jaynewstrom.concretesample.application;

import android.app.Application;

import com.jaynewstrom.concrete.Concrete;
import com.jaynewstrom.concrete.ConcreteWall;
import com.squareup.leakcanary.LeakCanary;

public final class ConcreteSampleApplication extends Application {

    private ConcreteWall<ApplicationComponent> foundation;

    @Override public void onCreate() {
        super.onCreate();
        LeakCanary.install(this);
        foundation = Concrete.pourFoundation(createApplicationComponent());
    }

    private ApplicationComponent createApplicationComponent() {
        return DaggerApplicationComponent.builder().applicationModule(new ApplicationModule(this)).build();
    }

    @Override public Object getSystemService(String name) {
        if (Concrete.isService(name)) {
            return foundation;
        }
        return super.getSystemService(name);
    }
}

File: concrete-sample/src/main/java/com/jaynewstrom/concretesample/application/ConcreteSampleApplication.kt
package com.jaynewstrom.concretesample.application

import android.app.Application

import com.jaynewstrom.concrete.Concrete
import com.jaynewstrom.concrete.ConcreteWall

class ConcreteSampleApplication : Application() {
    private lateinit var foundation: ConcreteWall<ApplicationComponent>

    override fun onCreate() {
        super.onCreate()
        foundation = Concrete.pourFoundation(createApplicationComponent())
    }

    private fun createApplicationComponent(): ApplicationComponent {
        return DaggerApplicationComponent.builder().applicationModule(ApplicationModule(this)).build()
    }

    override fun getSystemService(name: String): Any? {
        return if (Concrete.isService(name)) {
            foundation
        } else {
            super.getSystemService(name)
        }
    }
}
